### PR TITLE
treewide: use lib.optionalAttrs instead of 'then {}'

### PR DESCRIPTION
**Pull request by Stunkymonkey**
###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes (or backporting 23.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/lisp-modules-obsolete/quicklisp-to-nix-overrides.nix [hunk 1]
```diff
@@ -4,8 +4,7 @@ let
   skipBuildPhase = x: {
     overrides = y: ((x.overrides y) // { buildPhase = "true"; });
   };
-  multiOverride = l: x: if l == [] then {} else
-    ((builtins.head l) x) // (multiOverride (builtins.tail l) x);
+  multiOverride = l: x: lib.optionalAttrs (l != []) ((builtins.head l) x) // (multiOverride (builtins.tail l) x);
```
Comment (rhendric) [inline pkgs/development/lisp-modules-obsolete/quicklisp-to-nix-overrides.nix:7]: `lib` isn't in scope here.
```suggestion
  multiOverride = l: x: pkgs.lib.optionalAttrs (l != []) ((builtins.head l) x) // (multiOverride (builtins.tail l) x);
```

File: pkgs/development/coq-modules/tlc/default.nix [hunk 2]
```diff
@@ -23,9 +23,7 @@
     maintainers = [ maintainers.vbgl ];
   };
 }).overrideAttrs (x:
-  if lib.versionAtLeast x.version "20210316"
-  then {}
-  else {
+  lib.optionalAttrs (lib.versionAtLeast x.version "20210316") {
```
Comment (R-VdP) [inline pkgs/development/coq-modules/tlc/default.nix:26]: I think you need to invert the condition here, don't you?

**Review by drupol — APPROVED** [review]
LGTM

File: pkgs/development/coq-modules/tlc/default.nix [hunk 3]
```diff
@@ -23,9 +23,7 @@
     maintainers = [ maintainers.vbgl ];
   };
 }).overrideAttrs (x:
-  if lib.versionAtLeast x.version "20210316"
-  then {}
-  else {
+  lib.optionalAttrs (!(lib.versionAtLeast x.version "20210316")) {
```
Comment (rhendric) [inline pkgs/development/coq-modules/tlc/default.nix:26]:
```suggestion
  lib.optionalAttrs (lib.versionOlder x.version "20210316") {
```

Comment (drupol) [reply pkgs/development/coq-modules/tlc/default.nix:26]: Ah nice!

Comment (Stunkymonkey) [reply pkgs/development/coq-modules/tlc/default.nix]: that is much better thanks.

**Review by rhendric — APPROVED** [review]

**Comment by Stunkymonkey** [comment]
thanks for your good and thorough reviews :+1:

**Review by SuperSandro2000 — APPROVED** [review]

**Review by drupol — APPROVED** [review]

**Comment by vcunat** [comment]
Aaand again - broke evaluation and [is blocking channels](https://hydra.nixos.org/build/224144696).
```
$ nix eval -f. lispPackages.clsql.outPath
error: 'tail' called on an empty list
[...]
```

File: pkgs/development/lisp-modules-obsolete/quicklisp-to-nix-overrides.nix [hunk 4]
```diff
@@ -4,8 +4,7 @@ let
   skipBuildPhase = x: {
     overrides = y: ((x.overrides y) // { buildPhase = "true"; });
   };
-  multiOverride = l: x: if l == [] then {} else
-    ((builtins.head l) x) // (multiOverride (builtins.tail l) x);
+  multiOverride = l: x: pkgs.lib.optionalAttrs (l != []) ((builtins.head l) x) // (multiOverride (builtins.tail l) x);
```
Comment (vcunat) [inline pkgs/development/lisp-modules-obsolete/quicklisp-to-nix-overrides.nix:7]: I think missing parentheses here?

Comment (drupol) [reply pkgs/development/lisp-modules-obsolete/quicklisp-to-nix-overrides.nix:7]: Fixed in #237631

Comment (vcunat) [reply pkgs/development/lisp-modules-obsolete/quicklisp-to-nix-overrides.nix:7]: OK I guess, though I really suspect that that if-then-else version was more readable.  Also supported by getting several +1 reviews without noticing that the new code parses wrong.  (_copied this from chat, for reference_)

Comment (rhendric) [reply pkgs/development/lisp-modules-obsolete/quicklisp-to-nix-overrides.nix:7]: Possibly relevant to the proposal for `<|` in https://github.com/NixOS/rfcs/pull/148.

**Pull request merged**